cli/src/offline/blockhash_query.rs

```diff
@@ -0,0 +1,400 @@
+use super::*;
+
+#[derive(Debug, PartialEq)]
+pub enum Source {
+    Cluster,
+    NonceAccount(Pubkey),
+}
+
+impl Source {
+    pub fn get_blockhash_fee_calculator(
+        &self,
+        rpc_client: &RpcClient,
+    ) -> Result<(Hash, FeeCalculator), Box<dyn std::error::Error>> {
+        match self {
+            Self::Cluster => {
+                let res = rpc_client.get_recent_blockhash()?;
+                Ok(res)
+            }
+            Self::NonceAccount(ref pubkey) => {
+                let data = nonce::get_account(rpc_client, pubkey)
+                    .and_then(|ref a| nonce::data_from_account(a))?;
+                Ok((data.blockhash, data.fee_calculator))
+            }
+        }
+    }
+
+    pub fn get_fee_calculator(
+        &self,
+        rpc_client: &RpcClient,
+        blockhash: &Hash,
+    ) -> Result<Option<FeeCalculator>, Box<dyn std::error::Error>> {
+        match self {
+            Self::Cluster => {
+                let res = rpc_client.get_fee_calculator_for_blockhash(blockhash)?;
+                Ok(res)
+            }
+            Self::NonceAccount(ref pubkey) => {
+                let res = nonce::get_account(rpc_client, pubkey)
+                    .and_then(|ref a| nonce::data_from_account(a))
+                    .and_then(|d| {
+                        if d.blockhash == *blockhash {
+                            Ok(Some(d.fee_calculator))
+                        } else {
+                            Ok(None)
+                        }
+                    })?;
+                Ok(res)
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum BlockhashQuery {
+    None(Hash),
+    FeeCalculator(Source, Hash),
+    All(Source),
+}
+
+impl BlockhashQuery {
+    pub fn new(
```

Comment:
Fwiw, I think this constructor method is actually harder to grok in all those tests than just creating the BlockhashQuery as needed. One option would be to move this logic into `new_from_matches`, which I think is the only place you really need it, right?

Reply:
Ha! `new()` was broken out of `new_from_matches()` upon request _to facilitate testing_ during review when this feature originally came in.

I _do_ like being able to test the construction logic without building an `ArgMatches`, but could go either way in the actual test case instantiations. The main motivation of using `new()` is that the `Source` instantiation is a bit verbose

Reply:
Lol! 🤡 🎈 feel free to disregard my comment, then

Reply:
Came here to say the same thing, the constructor approach makes it hard to tell if the tests are setup correctly. Can we switch the tests back to the enum approach?

Reply:
Two votes wins.  First thing on the plate in the morning

Reply:
https://github.com/solana-labs/solana/pull/8750/commits/26e0edfb6086e9bde4d733024d404a40cad82cfc